java/src/jmri/jmrit/display/layoutEditor/ConnectivityUtil.java

```diff
@@ -716,12 +745,19 @@ public boolean layoutTurnoutHasRequiredSignals(LayoutTurnout t) {
      * @return a SignalHead facing away from or towards block depending on value
      *         of facing; may be null
      */
-    public SignalHead getSignalHeadAtAnchor(PositionablePoint p, Block block, boolean facing) {
+    @CheckReturnValue
+    @Nullable
+    public SignalHead getSignalHeadAtAnchor(@Nullable PositionablePoint p,
+            @Nullable Block block, boolean facing) {
         if ((p == null) || (block == null)) {
             log.error("null arguments in call to getSignalHeadAtAnchor");
```

Comment:
If they're @Nonnull then you wouldn't need to test them for null… right? I just assumed that since they were testing them then passing a null object was allowed. Is this not correct?

Reply:
This is a design question.  You get to decide where the null case should be handled.  Should it never be called with null? Then mark `@nonnull`, and CI (and eventually FindBugs) will force us to handle it in the calling places.  That's particularly useful if there's nothing meaningful that can be done inside the routine for the null case.  If you are coding an `if (p==null)` and can't figure out what to do next, that's a good reason to say `@nonnull`.

But if null is a useful or meaningful possibility (e.g. in a test for equality, it's not equal, no program), then handle it in the routine and don't use `@nonnull`.

Reply:
(Sorry, hit enter too soon)

If null is an _error_, then throwing IllegalArgumentException is another possibility. FindBugs will catch it at compile time (though we don't keep that clean) and the exception will force us improve the code if/when it gets hit.  Logging is a much weaker form of this.

But if null is sort-of-routine-but-not-great, then a _warning_ message might be appropriate. One hint of that is that it's returning null, which means that other code is already able to handle that null.

Individually, these are small things, but they're examples of the low-level, small-detail form of design improvements that make the code more robust over time.

Reply:
If you annotate a parameter as ```@Nonull``` then FindBugs will complain if you check it for null and do not throw an Exception. I would throw an NullPointerException (this is what the JDK does if an argument that should be non-null but isn't; it even provides java.lang.Objects.requireNonNull(...) methods that do this for you).

The sole use of the particular method immediately checks that the result of calling this isn't null, and treats a null result as an error, so having it catch an NPE is probably acceptable.